### ISSUE TITLE
Refactor notebooks to be able to run completely in Google's Colaboratory

### ISSUE THREAD

**Issue by thomasjpfan** [dnouri/skorch#316]
It could be beneficial to allow notebooks to  run completely in Colab. Sometimes the notebooks would have to be updated to be able to run in Colab. For example, here is an updated version of Transfer Learning that runs on Colab:

https://colab.research.google.com/github/thomasjpfan/skorch/blob/colab-basic-usage/notebooks/Transfer_Learning.ipynb

To run on a GPU, a user would need to navigate: `Runtime -> Change runtime type -> Hardware accelerator -> GPU -> Save`.

**Comment by taketwo** [comment]
Speaking of skorch tutorials/notebooks, I wanted o propose moving them to Sphinx-Gallery, the same as PyTorch. What do you think?

**Comment by benjamin-work** [comment]
In general, we should put more effort into facilitating learning and discovering more advanced features, so these ideas sound great.

> Sometimes the notebooks would have to be updated to be able to run in Colab

> Speaking of skorch tutorials/notebooks, I wanted o propose moving them to Sphinx-Gallery

Could you both quickly summarize the changes required for that?

**Comment by thomasjpfan** [comment]
For colab to work we need an additional cell to:

1. Install pytorch, skorch, and other dependencies.
2. Download data if needed.

In my example above, I needed to add the following cell at the top:

```bash
!pip install torch torchvision skorch pillow==4.1.1
!mkdir -p datasets
!wget -nc --no-check-certificate https://download.pytorch.org/tutorial/hymenoptera_data.zip -P datasets
!unzip -u datasets/hymenoptera_data.zip -d datasets
```

Tutorials that do not need to download data, just needs to `pip install` dependencies.

**Comment by taketwo** [comment]
Regarding Sphinx-Gallery, the tutorial code will need to be converted from Notebook format to normal Python scripts with prose written as rst markup in comments.

For example, here is the source code for one of PyTorch tutorials: [code](https://github.com/pytorch/tutorials/blob/master/beginner_source/blitz/cifar10_tutorial.py), and this is how it appears on their website: [html](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html).

The advantage of the Sphinx-Gallery approach is that the code gets re-executed during documentation generation, meaning that the outputs (both textual and images) are always up-to-date. Also you may note that at the bottom of the tutorial page there is a link to download the tutorial either as a Python script, or as a Notebook. Finally, using Sphinx-Gallery allows deeper integration (cross-linking) between tutorials and documentation.

**Comment by benjamin-work** [comment]
> For colab to work we need an additional cell to:

This looks like a minor change, especially if it is possible to detect colab from an environment variable and only execute the commands if this environment is set.

> Regarding Sphinx-Gallery, the tutorial code will need to be converted from Notebook format to normal Python scripts with prose written as rst markup in comments.

Okay, this looks like more work, since we don't want to maintain the notebook and the .py code separately. If we want to convert automatically, we will encounter some issues (OTOH):

* Things like magic commands are not automatically removed, breaking Python code.
* There are stray comments, like `ln[0]` for cell delimiters.
* Having to write the prose as commented rst clashes with the rendering in the notebook.

All those are issues that can be overcome but I wonder at what price. If there are finished solutions, though, we could use those (maybe [this](https://gist.github.com/chsasank/7218ca16f8d022e02a9c0deb94a310fe)?).

Regarding the advantages of using Sphinx-Gallery

> the code gets re-executed during documentation generation

This can also be a liability because it restricts notebooks to run in a reasonable time (including data download etc.). 

> the outputs (both textual and images) are always up-to-date

This is certainly nice to have.

> at the bottom of the tutorial page there is a link to download the tutorial either as a Python script, or as a Notebook

I think that more linking between docs and examples/notebooks could be beneficial but it does not necessarily need Sphinx-Gallery.

**Comment by taketwo** [comment]
> we don't want to maintain the notebook and the .py code separately

Sure. If we do this, the Sphinx-Gallery code would be the one and only original source, and corresponding Notebooks will be generated automatically together with the rest of the documentation.

> All those are issues that can be overcome but I wonder at what price. If there are finished solutions, though, we could use those

An automated solution like that one can be a great starting point. But most likely will require additional manual round of editing.

> This can also be a liability because it restricts notebooks to run in a reasonable time (including data download etc.).

Apparently this works for PyTorch, but I suppose they run documentation generation on their own servers. We most likely won't be able to execute all tutorials within 900s timeslot of Readthedocs. So this is a very valid concern indeed.

> I think that more linking between docs and examples/notebooks could be beneficial but it does not necessarily need Sphinx-Gallery

Sphinx-Gallery can automatically link functions _inside_ code blocks to their documentation. I think this is pretty cool and not possible to achieve with Notebooks.


To sum up, it would require some effort to convert to Sphinx-Gallery. Right now I don't want to commit to doing this. But that said, I suppose at some point when I have free time and mood I can do this.

So yes, it's going to require some effort. I'm not committing to do it (yet)

**Comment by benjamin-work** [comment]
> the Sphinx-Gallery code would be the one and only original source, and corresponding Notebooks will be generated automatically

> An automated solution like that one can be a great starting point. But most likely will require additional manual round of editing.

I believe this is crucial, if there isn't a fully automatic solution, the two documents will likely diverge more and more. It is already quite difficult to keep code, docstrings, docs, and notebooks/examples in sync.

> Apparently this works for PyTorch, but I suppose they run documentation generation on their own servers.

I believe they do, same with sklearn.

> Sphinx-Gallery can automatically link functions inside code blocks to their documentation

Yes, this is nice. Another big advantage over notebooks is version control.

> We most likely won't be able to execute all tutorials within 900s timeslot of Readthedocs

This is also my biggest concern from a technical side, especially since their servers won't be GPU-powered ;)

> To sum up, it would require some effort to convert to Sphinx-Gallery

From my perspective, jupyter notebooks cover 90% of the utility of Sphinx-Gallery and become increasingly more popular (rendered on github, Kaggle kernels and many cloud solutions). My guess is that Sphinx-Gallery will lose appeal over time, so the issue will be even less pressing in 6 months.

If we put the effort that would be required for this into improving the notebooks (keeping them up-to-date, adding more links to docs, etc.), skorch users will probably benefit more. But of course I can't tell for sure what users really want.
